### ISSUE TITLE
Interactions between _Expect and _Ignore

### ISSUE THREAD

**Issue by Miv13** [ThrowTheSwitch/Ceedling#286]
I have a problem (a few, actually) with making Expect and Ignore work as I need them to and I'm not quite sure whether what I'm trying to achieve is simply impossible or if I'm just missing something.

In a code for embedded memory that I use, some mocked functions funcA, funcB and funcC are all called in my tested function, in this specific order. Moreover, I have a loop which might called them more times (in the same order), but I don't care how many times. For some reason (probably related to the rest of my code), if I Ignore all 3 functions, the test just runs forever (but that's probably just my problem).

What I wanted to do, was to make the test check if all the functions were called AT LEAST once (before the loop) and not care whether they were called any more times. I hoped that using Expect and then Ignore would make that possible, but if I specify them like that, the Expects get, well, ignored. If I use:

funcC_Expect();
funcA_Ignore();
funcB_Ignore();

the test fails, because funcC was called more times than expected (which is as it should be, because of the loop). But here comes the fun part - despite the Expects being overwritten by Ignores, if I use:

funcA_Expect();
funcB_Expect();
funcC_Expect();
funcA_Ignore();
funcB_Ignore();

then the test fails as well, but for a different reason: funcC being called earlier than expected. This would indicate that using Expect and then Ignore IS different than just using Ignore, only not in the way that would be useful...

Furthermore, if I use:

funcA_Ignore();
funcB_Ignore();
funcC_Ignore();
funcA_Expect();
funcB_Expect();
funcC_Expect();

then the test fails because of funcA being called later than expected.

If I just leave:

funcA_Ignore();
funcB_Ignore();
funcC_Ignore();
funcC_Expect();

then it seems that the Expect actually overwrites Ignore and the failure reason is funcC being called more times than expected. Why doesn't it work if I overwrite all of them?

Is such behaviour intended? And is there any way to check whether a mocked function was called AT LEAST once? I have to check it, because all this happens inside an if statement, so simply ignoring these functions won't do. I also cannot disable strict ordering, as it is quite important for my tests.

I couldn't find any issues (or solutions) close enough to what I'm facing. If you need any more details or explanations, please let me know - I know I tend to make things quite messy when I try too hard to explain stuff.

**Comment by mvandervoord** [comment]
This is a really good question and I appreciate you putting the time into explaining your situation so clearly. Examples like this are really helpful!

First, the core behavior you are looking for does not exist at this time, but it IS a planned behavior. We hope to include options for Expecting n+ or n- calls in the future. It would be REALLY helpful for situations like yours.

Second, An Expect DOES override an Ignore. It's as if the Ignore never happened when you do that. The reason it didn't seem to work in your second to last example (I believe) is that you've run into the global ordering problem. If you disable global ordering, it's no longer going to care which order you call funcA, funcB, etc in... it's only going to care that you called each the correct number of times by the time you are done. (which doesn't solve your root problem, as explained above, but does explain this extra quirk).

For now, the most complete workaround is to use the stub_with_callback plugin and set yourself a flag to let you know it was called at least once, and use an assert at the end of your test to check that flag. It's not pretty, but it's what CMock would do under the hood once it gets this feature.

Sorry about the inconvenience!

**Comment by Miv13** [comment]
I'm afraid I still don't understand why in my second to last example funcA is said to be called later than expected. It is supposed to be the first function called in this test - and it is. If I only use:

funcA_Expect();
funcB_Expect();
funcC_Expect();

then there is no ordering problem, just the standard "Called more times than expected." (which means that i went through the first use of each function without a fail).
This would indicate that, while Ignore overwrites Expect and Expect overwrites Ignore, the first (overwritten) statement still influences the test in some way, causing an ordering failure. While just solving this would not yet solve my problem, I think it would be good to look into it, as it might cause problems to others. This also prevents people from setting a 'default' behaviour (e.g. Ignore) in setUp that you could overwrite in specific tests - which should work fine, if the behaviour specified in setUp wouldn't influence the tests despite later Expect statements (I checked this and also got the "Called later than expected" fail).

Could you tell me where do I define the Callback function? The CMock doc (https://github.com/ThrowTheSwitch/CMock/blob/master/docs/CMock_Summary.md) does not seem to explain this clearly enough (at least for me).

When could this new "Expecting n+ or n- calls" functionality be implemented (days/weeks/months)?

Thanks for the quick answer, really appreciate it!

**Comment by mvandervoord** [comment]
Ah, I missed that detail. That seems like an additional bug. It's possible the global ordering counter is being affected by ignores as well, though I'm not sure why that would be. Clearly the global ordering feature and the ignore feature aren't cooperating somehow.

For a callback, you want to add the callback function to your test file... for example something like this:

```
int has_funcA_been_called;

void funcA_callback(void) {
    has_funcA_been_called = 1;
}

void setUp(void) {
    has_funcA_been_called = 0;
}

void test_ThatFuncAGetsCalled(void) {

    funcA_StubWithCallback(funcA_callback); // This could alternatively go in setUp if you want to use it in ALL your tests in this file

    function_to_test();

    TEST_ASSERT_TRUE_MESSAGE(has_funcA_been_called, "Expected Function A To Be Called But It Was Never Called");
}
```

**Comment by Miv13** [comment]
Thanks a lot!

This should suffice until you get the new functionality going. Should I leave the issue open until you add it and fix the ignore/ordering problem?

**Comment by Miv13** [comment]
One more question about the StubWithCallback - do I HAVE TO specify the parameters it will be called with? Or is there an equivalent for ExpectAnyArgs? I kind of forgot that my actual functions aren't as simple as the ones I gave you as examples...

**Comment by mvandervoord** [comment]
Yes, leaving this open for me please. It's a good reminder I have work to do. ;)

For StubWithCallback: you specify that you want a callback function to be called with the StubWithCallback call... From there on, your callback is in charge of return values, validating arguments, or anything else you would like it to accomplish. There IS an option called :callback_after_arg_check which can be set to TRUE if you want it to require your normal Expect calls each time... but given your description, that's not what you want here.

**Comment by Miv13** [comment]
If I may have an additional (somewhat connected) question, for which I don't want to open another issue:

Is it possible to specify additional parameters for the _MESSAGE in test assertions, like in printf function? My test assertions for the above mentioned problem would look more or less like this:

TEST_ASSERT_TRUE_MESSAGE(number_funcA_called >= number_funcA_expected, ("FuncA was called lest times than expected (at least %d).\n", number_funcA_expected));

This way I can easily control the number of times each function is expected in each test as well as make sure that I don't accidentally output a wrong expectation in the failure message. For example: I copied the test assertion from a function when the expected number was 1 into a function where it's supposed to be 2 and changed the number in test, but forgot to change it in the message. Then if I run the test and the function will only be called once, it will output a message that it was supposed to be called once (while it should really be called twice), which it actually was - so I will be confused and try to find the reason why wasn't it called even once (I'll check why the first appearance of the function wasn't triggered) instead of looking for the actual reason for test failure.

Right now this generates an error, I assume it's due to the incorrect argument corresponding to the MESSAGE part. Is there any way to implement this in my tests? (I know I could do this with printf etc.. but I guess the whole point of using a test framework is so that I don't have to do things manually like that)

**Comment by Miv13** [comment]
OK, I have additional note to the issue: Expect used after StubWithCallback seems to be causing similar ordering problems as Ignore. If I use:

funcA_StubWithCallback(funcA_callback);
funcA_Expect(123); // the value with which it's called is different
funcD_Expect();

the funcA_Expect(123) is mostly ignored (i.e. there is no failure due to unexpected argument value), but the test fails because funcD was "called earlier than expected". So it seems that using funcA_Expect (I know, it shouldn't be used like that in the first place) infuences the expected global order, causing funcD to be expected as the second function called and not the first. Thought you might be interested in this, as it may help you find the problem or at least point out that it isn't just the case with Ignore.

Also, another side question: is it impossible to ignore just SOME arguments passed into an expected function? I have a function which is called with 2 arguments - one of them must have a specified value and the other is an address of a variable initiated inside of the function (so I can't really know its address beforehand, can I?). I'd like to check the first one and ignore the second, as it always gives me the "unexpected argument value" fail.

I hope I'm not too much of a bother... :P

**Comment by mvandervoord** [comment]
You can't specify printf-style formatting in assertions, no. Many embedded processors can't handle printf and so we purposefully avoid all that overhead.

If you use StubWithCallback it will use the stub for everything unless you specify :callback_after_arg_check as TRUE. In that case, it should require both expects and stubs. It doesn't surprise me that there would be a similar global ordering issue using both. It's a rare corner case that you would want both and it's probably the same root problem as the Ignore function.

To ignore arguments, you have a couple choices. For both, you'll need to enable extra plugins in your CMock config.

**:expect_any_args**

This creates another function which is like funcA_ExpectAndArgs and funcB_ExpectAndArgsAndReturn. They work just like Expect except that you don't specify arguments. It will ignore all of them and only cares that you called the function when you should.

**:ignore_arg** 

This creates an additional function for each argument to allow you to ignore it. You specify this after your Expect call. So let's say you have function being mocked like this:

```
void funcK(int uno, int* dos);
```

Then in a test you could have it supply `uno` and ignore `dos`... you'll still give `dos` something for now, but then tell it to be ignored. Like so:

```
funcK_Expect(1, NULL);
funcK_IgnoreArg_dos();
```

**Comment by Miv13** [comment]
So one can only specify a 'rigid' string as the message? I was kind of expecting that, but thanks for the confirmation.

I've been using the 'expect_any_args' plugin so far, I'm glad there is a more precise tool for this.

Thank you for having patience with me...

**Comment by mvandervoord** [comment]
Yes, only rigid strings. If you absolutely need formatting, there is always snprintf and then the message.

However, you can get the message you want by using a more precise assertion:

```
TEST_ASSERT_LESS_OR_EQUAL_MESSAGE(number_funcA_expected, number_funcA_called, "FuncA called wrong number of times");
```

**Comment by Miv13** [comment]
Seems like my problem is impossible to resolve at the moment, even with StubWithCallback. 

1. To be able to check the arguments despite StubWithCallback, I must set ':callback_after_arg_check: TRUE', right? (which is set globally, so affects all tests)
2. This in turn makes it necessary to specify Expect for every function call. And avoiding such situation was the very reason I used StubWithCallback.

This means that I cannot check both the arguments and min. number of function calls. Or am I missing something?

**Comment by mvandervoord** [comment]
I would NOT enable :callback_after_arg_check and would leave the default behavior running.

Inside your callback, I would check the argument(s) you care about. There's no reason you can't have as many assertions as you want in your callback function.

**Comment by Miv13** [comment]
It's me again...

Is there a way to access variables initialized inside of the tested function within the callback? (e.g. for assertions, since some of those variables are passed as arguments to my mocked functions)
I wasn't able to find any ceedling/CMock examples that would show assertion of argument values inside a callback.

**Comment by mvandervoord** [comment]
If you're mocking a function which accepts arguments, then those arguments get passed along to your callback. For example, if you have a function like this that you are mocking:

```
int Doubler(int value);
```

and your function you're testing looks something like this:

```
int FetchSillyNumber(int times) 
{
    int i;
    int sum = 7;

    for (i=0; i < times; i++)
         sum = Doubler(sum);

    return sum;
}
```

Then you might want to write a test like this one, if you were going to use a Stub:

```
static int CallCounter = 0;

void setUp(void)
{
    CallCounter = 0;
}

int DoublerCallback(int val)
{
    if (CallCounter == 0) {
        TEST_ASSERT_EQUAL_INT_MESSAGE(7, val, "Received Incorrect Value");
    }
    TEST_ASSERT_LESS_THAN_INT_MESSAGE(3, ++CallCounter);
    return val*2;
}

void test_FetchSillyNumber_should_CallDoublerLessThanThreeTimes(void)
{
    Doubler_StubWIthCallback(DoublerCallback);

    TEST_ASSERT_EQUAL(28, FetchSillyNumber());
}
```

Does that make sense?

**Comment by Miv13** [comment]
Other than the fact that you didn't pass any message in TEST_ASSERT_LESS_THAN_INT_MESSAGE and didn't pass '2' to FetchSillyNumber() in the last assertion - yes, I think so... :P
Or is the latter done on purpose? Does the "TEST_ASSERT_LESS_THAN_INT_MESSAGE(3, ++CallCounter);" somehow ensure that FetchSillyNumber() will only be called twice? But even in that case it would still give an error because of FetchSillyNumber() being called with no argument, right?

I'll have to create something similar myself and play with it a bit to make sure I get it right...

Another small thing: as far as I understand, the only way to control the arguments passed to the mocked function with each call (if they're different each time) is with "if (CallCounter == sth)"? You can only use the "Doubler_StubWIthCallback(DoublerCallback);" once in a test? (and otherwise it will simply be overwritten, not called differently as is the case with Expect?)

**Comment by mvandervoord** [comment]
Whoops. No no no, you're right. That's what I get for just making up a weird example in a text editor and not actually running it. Both of those things are wrong. :)

Your note at the end is correct. A call to StubWithCallback SETS the callback function. That callback function will be used EVERY time that function was to be called IN THAT TEST. So if you have to differentiate multiple calls in your test, it's left to you to do so.

One thing to note: I had forgotten that :callback_include_count: is enabled by default. This is going to help you. It means that for every callback, you will be getting an additional integer added AFTER the other arguments (if they exist). This number is the number of times that this function has been called... so you don't need to track it on your own as we have been above. Instead, our callback would look like this:

```
int DoublerCallback(int val, int num_calls)
{
    if (num_calls == 0) {
        TEST_ASSERT_EQUAL_INT_MESSAGE(7, val, "Received Incorrect Value");
    }
    TEST_ASSERT_LESS_THAN_INT(3, num_calls);
    return val*2;
}
```

All that nonsense with CallCounter can just go away, then, because you know that CMock is counting for you.

People use const arrays often for this sort of testing. For example, I could rewrite callback to be even better:

```
int DoublerCallback(int val, int num_calls)
{
   const int inputs[] = { 7, 14, 28 };
   const int outputs[] = { 14, 28, 56 };

   TEST_ASSERT_LESS_THAN_INT(3, num_calls);
   TEST_ASSERT_EQUAL_INT_MESSAGE(inputs[num_calls], val, "Received Incorrect Value");

   return outputs[num_calls];
}
```

That cleans things up to make it more clear what we are testing for, and it doesn't have us duplicating real code in our stub.

Finally, I should put out a reminder that MOST of the time, we don't want stubs/callbacks. These are our last resort, really, for when we're looking for unknown numbers of calls or we need to do something unique with the data. Most of the time, we stick with the more straightforward Expect calls and friends.

**Comment by Miv13** [comment]
This looks like a pretty decent solution, thank you.

One more (hopefully last) thing - is there any way I could access variables passed to the tested function from within the stub? I have a situation like this:

```
tested_function(struct_arg, len, buf){

struct_type STRUCT_X;
int temp_var;
int temp_arr[n];

temp_var = struct_arg.value << an_int | struct_arg.offset;
temp_arr[0] = (temp_var & const_hex) >> 16; //sth sth
// similar for rest of the array

WEIRD_MACRO_ASSIGNING_STUFF_TO_STRUCT_X(temp_arr, n, len, buf);

mocked_function(some_int, &STRUCT_X, some_other_int);
```

I use StubWithCallback on the mocked_function which allows me to test most of the stuff I need. However, I'm not sure how to handle testing values of struct_arg, since it is only passed to the tested _function and used indirectly to derive other variables which are then passed to the mocked_function.
I compare the other arguments (len, buf) directly to the fields of STRUCT_X, but since struct_arg isn't directly related to it, I cannot simply do sth like:

`TEST_ASSERT_EQUAL(len, STRUCT_X->len_field);`

I would have to somehow revert from STRUCT_X->some_fields to temp_arr to temp_var and then to struct_arg, which I'm not sure I can do with binary compares etc. happening in between.

Any ideas?
